src/lib.rs

```diff
@@ -290,942 +143,48 @@ impl DiversifierIndex {
-
-            assert_eq!(xsk.expsk.ovk.0, tv.ovk);
-            assert_eq!(xsk.dk.0, tv.dk);
-            assert_eq!(xsk.chain_code.0, tv.c);
-
-            let mut ser = vec![];
-            xsk.write(&mut ser).unwrap();
-            assert_eq!(&ser[..], &tv.xsk.unwrap()[..]);
-        }
-
-        for j in 0..xfvks.len() {
-            let xfvk = &xfvks[j];
-            let tv = &test_vectors[j];
-
-            let mut buf = [0; 32];
-            xfvk.fvk.vk.ak.write(&mut buf[..]).unwrap();
-            assert_eq!(buf, tv.ak);
-            xfvk.fvk.vk.nk.write(&mut buf[..]).unwrap();
-            assert_eq!(buf, tv.nk);
-
-            assert_eq!(xfvk.fvk.ovk.0, tv.ovk);
-            assert_eq!(xfvk.dk.0, tv.dk);
-            assert_eq!(xfvk.chain_code.0, tv.c);
-
-            xfvk.fvk
-                .vk
-                .ivk()
-                .into_repr()
-                .write_le(&mut buf[..])
-                .unwrap();
-            assert_eq!(buf, tv.ivk);
-
-            let mut ser = vec![];
-            xfvk.write(&mut ser).unwrap();
-            assert_eq!(&ser[..], &tv.xfvk[..]);
-            assert_eq!(xfvk.fvk.fingerprint().0, tv.fp);
-
-            // d0
-            let mut di = DiversifierIndex::new();
-            match xfvk.dk.diversifier(di) {
-                Ok((l, d)) if l == di => assert_eq!(d.0, tv.d0.unwrap()),
-                Ok((_, _)) => assert!(tv.d0.is_none()),
-                Err(_) => panic!(),
-            }
-
-            // d1
-            di.increment().unwrap();
-            match xfvk.dk.diversifier(di) {
-                Ok((l, d)) if l == di => assert_eq!(d.0, tv.d1.unwrap()),
-                Ok((_, _)) => assert!(tv.d1.is_none()),
-                Err(_) => panic!(),
-            }
-
-            // d2
-            di.increment().unwrap();
-            match xfvk.dk.diversifier(di) {
-                Ok((l, d)) if l == di => assert_eq!(d.0, tv.d2.unwrap()),
-                Ok((_, _)) => assert!(tv.d2.is_none()),
-                Err(_) => panic!(),
-            }
-
-            // dmax
-            let dmax = DiversifierIndex([0xff; 11]);
-            match xfvk.dk.diversifier(dmax) {
-                Ok((l, d)) if l == dmax => assert_eq!(d.0, tv.dmax.unwrap()),
-                Ok((_, _)) => panic!(),
-                Err(_) => assert!(tv.dmax.is_none()),
-            }
-        }
+    fn diversifier_index_to_u32() {
+        let two = DiversifierIndex([
+            0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        ]);
+        assert_eq!(u32::try_from(two), Ok(2));
+
+        let max_u32 = DiversifierIndex([
+            0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        ]);
+        assert_eq!(u32::try_from(max_u32), Ok(u32::MAX));
+
+        let too_big = DiversifierIndex([
+            0xff, 0xff, 0xff, 0xff, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        ]);
+        assert_matches!(u32::try_from(too_big), Err(_));
     }
```

Comment:
Consider adding more tests, in particular for `DiversifierIndex::increment`.

Reply:
Opened #8.